src/vmm/src/devices/virtio/vhost_user.rs

```diff
@@ -257,3 +464,517 @@ impl VhostUserHandle {
+                tmp_socket_path.as_ptr().cast(),
+                socket_addr.sun_path.as_mut_ptr(),
+                tmp_socket_path.as_bytes().len(),
+            );
+
+            let bind = libc::bind(
+                socketfd,
+                (&socket_addr as *const libc::sockaddr_un).cast(),
+                std::mem::size_of::<libc::sockaddr_un>() as u32,
+            );
+            if bind < 0 {
+                panic!("Cannot bind socket");
+            }
+
+            let listen = libc::listen(socketfd, 1);
+            if listen < 0 {
+                panic!("Cannot listen on socket");
+            }
+        }
+
+        let vuh = VhostUserHandleImpl::<MockMaster>::new(&tmp_socket_path, max_queue_num).unwrap();
+        assert_eq!(
+            vuh.vu
+                .sock
+                .peer_addr()
+                .unwrap()
+                .as_pathname()
+                .unwrap()
+                .to_str()
+                .unwrap(),
+            &tmp_socket_path,
+        );
+        assert_eq!(vuh.vu.max_queue_num, max_queue_num);
+        assert!(unsafe { *vuh.vu.is_owner.get() });
+    }
+
+    #[test]
+    fn test_set_features() {
+        struct MockMaster {
+            features: std::cell::UnsafeCell<u64>,
+        }
+
+        impl VhostUserHandleBackend for MockMaster {
+            fn set_features(&self, features: u64) -> Result<(), vhost::Error> {
+                unsafe { *self.features.get() = features };
+                Ok(())
+            }
+        }
+
+        let vuh = VhostUserHandleImpl {
+            vu: MockMaster { features: 0.into() },
+            socket_path: "".to_string(),
+        };
+
+        vuh.set_features(0x69).unwrap();
+        assert_eq!(unsafe { *vuh.vu.features.get() }, 0x69);
+    }
+
+    #[test]
+    fn test_set_protocol_features() {
+        struct MockMaster {
+            protocol_features: VhostUserProtocolFeatures,
+            hdr_flags: std::cell::UnsafeCell<VhostUserHeaderFlag>,
+        }
+
+        impl VhostUserHandleBackend for MockMaster {
+            fn set_hdr_flags(&self, flags: VhostUserHeaderFlag) {
+                unsafe { *self.hdr_flags.get() = flags };
+            }
+
+            fn set_protocol_features(
+                &mut self,
+                features: VhostUserProtocolFeatures,
+            ) -> Result<(), vhost::Error> {
+                self.protocol_features = features;
+                Ok(())
+            }
+        }
+
+        let mut vuh = VhostUserHandleImpl {
+            vu: MockMaster {
+                protocol_features: VhostUserProtocolFeatures::empty(),
+                hdr_flags: std::cell::UnsafeCell::new(VhostUserHeaderFlag::empty()),
+            },
+            socket_path: "".to_string(),
+        };
+
+        // No protocol features are set if acked_features do not have PROTOCOL_FEATURES bit
+        let acked_features = 0;
+        let acked_protocol_features = VhostUserProtocolFeatures::all();
```

Comment:
can we add a case where `acked_protocol_features` is None?

Reply:
added